File: java-module/src/main/java/com/pfefferminzia/kfz/service/ValidationService.java
package com.pfefferminzia.kfz.service;

import com.pfefferminzia.kfz.common.PlausiException;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validation;
import jakarta.validation.Validator;
import jakarta.validation.ValidatorFactory;

import java.util.List;
import java.util.Set;

public class ValidationService {
    
    private static final Validator validator;
    
    static {
        try (ValidatorFactory factory = Validation.buildDefaultValidatorFactory()) {
            validator = factory.getValidator();
        }
    }

    public static <T> void validate(T object) {
        Set<ConstraintViolation<T>> violations = validator.validate(object);
        
        if (!violations.isEmpty()) {
            List<String> fehler = violations.stream()
                .map(ValidationService::formatFehler)
                .toList();
            
            throw new PlausiException(fehler);
        }
    }
    
    private static <T> String formatFehler(ConstraintViolation<T> violation) {
        String propertyPath = violation.getPropertyPath().toString();
        String message = violation.getMessage();
        
        // Formatierung für verschachtelte Objekte
        if (propertyPath.contains(".")) {
            String[] parts = propertyPath.split("\\.");
            String objectName = parts[0];
            String fieldName = parts[1];
            
            return switch (objectName) {
                case "haftpflicht" -> formatHaftpflichtFehler(fieldName, message);
                case "kasko" -> formatKaskoFehler(fieldName, message);
                default -> message;
            };
        }
        
        return message;
    }
    
    private static String formatHaftpflichtFehler(String field, String message) {
        return switch (field) {
            case "praemie" -> "Prämie für die Haftpflicht muss positiv sein";
            case "rabatt" -> {
                if (message.contains("negativ")) {
                    yield "Rabatt für die Haftpflicht darf nicht negativ sein";
                } else {
                    yield "Rabatt für die Haftpflicht darf nicht größer 99 sein";
                }
            }
            default -> message;
        };
    }
    
    private static String formatKaskoFehler(String field, String message) {
        return switch (field) {
            case "praemie" -> "Prämie für die Kasko muss positiv sein";
            case "rabatt" -> {
                if (message.contains("negativ")) {
                    yield "Rabatt für die Kasko darf nicht negativ sein";
                } else {
                    yield "Rabatt für die Kasko darf nicht größer 99 sein";
                }
            }
            default -> message;
        };
    }
}

File: dotnet-module/KfzVersicherung/Service/ValidationService.cs
using System.ComponentModel.DataAnnotations;
using DotnetModule.Common;

namespace DotnetModule.Service;

public static class ValidationService
{
    public static void Validate<T>(T obj, string? propertyPrefix = null)
    {
        var context = new ValidationContext(obj);
        var results = new List<ValidationResult>();

        if (!Validator.TryValidateObject(obj, context, results, true))
        {
            var fehler = results
                .Select(r => FormatFehler(r, propertyPrefix))
                .ToList();

            throw new PlausiException(fehler);
        }
    }

    private static string FormatFehler(ValidationResult result, string? propertyPrefix)
    {
        var message = result.ErrorMessage ?? "Validierungsfehler";
        
        if (propertyPrefix != null)
        {
            return propertyPrefix switch
            {
                "haftpflicht" => FormatHaftpflichtFehler(message),
                "kasko" => FormatKaskoFehler(message),
                _ => message
            };
        }

        return message;
    }

    private static string FormatHaftpflichtFehler(string message)
    {
        if (message.Contains("Prämie"))
            return "Prämie für die Haftpflicht muss positiv sein";
        if (message.Contains("Rabatt") && message.Contains("negativ"))
            return "Rabatt für die Haftpflicht darf nicht negativ sein";
        if (message.Contains("Rabatt"))
            return "Rabatt für die Haftpflicht darf nicht größer 99 sein";
        
        return message;
    }

    private static string FormatKaskoFehler(string message)
    {
        if (message.Contains("Prämie"))
            return "Prämie für die Kasko muss positiv sein";
        if (message.Contains("Rabatt") && message.Contains("negativ"))
            return "Rabatt für die Kasko darf nicht negativ sein";
        if (message.Contains("Rabatt"))
            return "Rabatt für die Kasko darf nicht größer 99 sein";
        
        return message;
    }
}
